packages/wallet/test/integration/buildTx.test.ts

```diff
@@ -0,0 +1,143 @@
+  expect(txOut.isValid).toBe(true);
+}
+
+describe('buildTx', () => {
+  let wallet: ObservableWallet;
+
+  beforeEach(async () => {
+    ({ wallet } = await createWallet());
+  });
+
+  afterEach(() => wallet.shutdown());
+
+  it('can be used to build, sign and submit a tx', async () => {
+    const tx = await buildTx(wallet).addOutput(mocks.utxo[0][1]).build();
+    if (tx.isValid) {
+      const signedTx = await tx.sign();
+      await signedTx.submit();
+    } else {
+      expect(tx.errors.length).toBeGreaterThan(0);
+      throw new Error('Invalid tx');
+    }
+  });
+
+  it('can add outputs one by one', async () => {
+    const tx = await buildTx(wallet).addOutput(mocks.utxo[0][1]).addOutput(mocks.utxo[1][1]).build();
+    assertTxIsValid(tx);
+    // 2 outputs + change output
+    expect(tx.body.outputs.length).toBe(3);
+  });
+
+  it('can remove outputs and build', async () => {
+    const txBuilder = buildTx(wallet);
+    const txTwoOutputs = txBuilder.addOutput(mocks.utxo[0][1]).addOutput(mocks.utxo[1][1]);
+    const tx = await txTwoOutputs.removeOutput(mocks.utxo[0][1]).build();
+    assertTxIsValid(tx);
+    expect(tx.body.outputs.length).toBe(2);
+  });
+
+  it('builds transactions that are not modified by subsequent builder changes', async () => {
+    const txBuilder = buildTx(wallet);
+    const tx = await txBuilder.addOutput(mocks.utxo[0][1]).build();
+    await txBuilder.addOutput(mocks.utxo[1][1]).build();
+
+    assertTxIsValid(tx);
+    expect(tx.body.outputs.length).toBe(2);
+  });
+
+  it('validates multiple outputs before building', async () => {
+    const coinMissingValidation: OutputValidation = {
+      coinMissing: 1n,
+      minimumCoin: 2n,
+      tokenBundleSizeExceedsLimit: false
+    };
+
+    const bundleSizeValidation: OutputValidation = {
+      coinMissing: 0n,
+      minimumCoin: 2n,
+      tokenBundleSizeExceedsLimit: true
+    };
+
+    const output1 = mocks.utxo[0][1];
+    const output2 = mocks.utxo[1][1];
+
+    const mockValidator: OutputValidator = {
+      validateOutput: jest.fn(),
+      validateOutputs: jest.fn(() =>
+        Promise.resolve(
+          new Map([
+            [output1, coinMissingValidation],
+            [output2, bundleSizeValidation]
+          ])
+        )
+      ),
+      validateValue: jest.fn(),
+      validateValues: jest.fn()
+    };
+    const tx = await buildTx(wallet, mockValidator).addOutput(output1).addOutput(output2).build();
+
+    expect(tx.isValid).toBeFalsy();
+    const [error1, error2] = (!tx.isValid && tx.errors) || [];
+
+    expect(error1 instanceof OutputValidationError).toBeTruthy();
+    if (error1 instanceof OutputValidationError) {
+      expect(error1.failure === TxOutputFailure.coinMissing);
+    }
+
+    expect(error2 instanceof OutputValidationError).toBeTruthy();
+    if (error2 instanceof OutputValidationError) {
+      expect(error2.failure === TxOutputFailure.tokenBundleSizeExceedsLimit);
+    }
+  });
+
+  describe('buildOutput', () => {
```

Comment:
It would be good to have a `describe` block for each method up the top of the suite, to ensure the precise behavior for each is covered, as this test suite will become quite large. Tests to demonstrate sequences can then be positioned below